### PR TITLE
Make ModelWrapValidator protocols generic

### PR DESCRIPTION
**Pull request by dmontagu**
Closes https://github.com/pydantic/pydantic/issues/7152

@samuelcolvin @adriangb is there any good way to add a test for this? Do we care enough to bother?

### CODE REVIEW

**Review by adriangb — APPROVED** [review]

File: pydantic/functional_validators.py [hunk 1]
```diff
@@ -423,7 +423,7 @@ def __call__(  # noqa: D102
 ModelAfterValidator = Callable[[_ModelType, _core_schema.ValidationInfo], _ModelType]
 """A `@model_validator` decorated function signature. This is used when `mode='after'`."""
 
-_AnyModelWrapValidator = Union[ModelWrapValidator, ModelWrapValidatorWithoutInfo]
+_AnyModelWrapValidator = Union[ModelWrapValidator[_ModelType], ModelWrapValidatorWithoutInfo[_ModelType]]
```
Comment (adriangb) [inline pydantic/functional_validators.py:426]: Doesn't this need to be parametrized later?

**Comment by adriangb** [comment]
Unfortunately adding tests for these things is not easy. If the original example now works locally with mypy and pyright let's merge this.

File: pydantic/functional_validators.py (hunk 1 above)
Comment (object-Object) [reply pydantic/functional_validators.py:426]: From my testing, adding the `_ModelType` parameter to the overload return type on line 435 seems to give the right results for a correct and incorrect validator (and the `after` overload does the same thing already):
```py
) -> Callable[[_AnyModelWrapValidator[_ModelType]], _decorators.PydanticDescriptorProxy[_decorators.ModelValidatorDecoratorInfo]]:
```
That appears to be the only place where these types are used.

**Review by samuelcolvin — APPROVED** [review]
I'd agree with @adriangb.

LGTM.

**Pull request merged**